### DISCUSSION TITLE
Constant 'Folding' Attribute

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/csharplang#5347] [Language Ideas]
I was just remarking (on a different issue) that it would be neat if the compiler could implement [constant folding](https://en.wikipedia.org/wiki/Constant_folding#:~:text=Constant%20folding%20is%20the%20process,are%20known%20at%20compile%20time.) with method calls and allow this kind of thing (currently illegal):

```cs
private const double Rate = Math.Pow(2, 6);
```

We (but not the compiler) "know" that that can be evaluated at compile time and the method call replaced with a constant (the result of the `Pow` call).

To support this we might need a new attribute `[FoldableResultAttribute]` that code authors can add to methods like this so that the compiler can safely replace such method calls with constant args.

The rule would be IF the args to a static method are constant AND the method is decorated with `[FoldableResultAttribute]` then and only then, is it safe to replace the method call with a constant result evaluated at compile time, enabling the above example to compile cleanly.

**Comment by CyrusNajmabadi** [comment]
Dupe of: https://github.com/dotnet/csharplang/discussions/1028

**Comment by deeprobin** [comment]
I think that will massively increase the build time... 
Especially if you use NuGet packages, which uses a lot of them. 

This is for example the case with Rust, that as soon as Procedural Macros are used, usually the build time at least doubles.

**Reply by BreyerW** [reply]
This could probably be alleviated partially by offering mechanism that would tell the compiler to ignore attribute when doing debug builds (or any builds  if it were switch in config file or somesuch)

**Comment by AartBluestoke** [comment]
would it be possible for a [RunByCompiler] attribute to work?
- this attribute instructs the compiler to attempt execute the next line of code, which initializes a const variable, at compile time, 

- it should be reasonably easy to determine that the execution is isolated; 
-- the code results in a const type (value, or string; perhaps some structs) 
-- the code does not read or write any static variables (should be the only way for data to 'escape')
--- in this mode, static memory access type is simply an error
-- the code does no IO (disk, screen, network, keyboard, etc ...) (we might be able to say 'no system calls' here)
-- all created objects can be garbage collected at the end of the execution -a post-condition escape analysis - if you can't GC everything just created, something has escaped, and so we need to abort the attempt to compile-time this function.
note, this is 'as it runs' check, not a syntax analysis, executing stuff then checking to see if the post-conditions are met (of no escape).

**Reply by HaloFour** [reply]
See: https://github.com/dotnet/csharplang/discussions/2379
